### PR TITLE
feat(helper): Adding one helper per node

### PR DESCRIPTION
**Pull request by ispeakc0de**
Signed-off-by: Shubham Chaudhary <shubham.chaudhary@harness.io>

**What this PR does / why we need it**:

## PR Items
- Experiment pod to launch a single helper pod per node for all target pods
- Updated the helper pods label filtering logic to support only single labels for both serial and parallel modes
- Using generate name for helper pod. Filtering pods based on labels instead of pod's name

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Checklist:**
-   [ ] Fixes #<issue number>
-   [ ] PR messages has document related information
-   [ ] Labelled this PR & related issue with `breaking-changes` tag
-   [ ] PR messages has breaking changes related information
-   [ ] Labelled this PR & related issue with `requires-upgrade` tag
-   [ ] PR messages has upgrade related information
-   [ ] Commit has unit tests
-   [ ] Commit has integration tests
-   [ ] E2E run Required for the changes

### CODE REVIEW

File: chaoslib/litmus/container-kill/helper/container-kill.go [hunk 1]
```diff
@@ -49,50 +54,66 @@ func Helper(clients clients.ClientSets) {
...
+		td := targetDetails{
+			Name:            target[0],
+			Namespace:       target[1],
+			TargetContainer: target[2],
+		}
+		targets = append(targets, td)
+	}
+
+	if err := killIterations(targets, experimentsDetails, clients, eventsDetails, chaosDetails, resultDetails); err != nil {
+		return err
```
Comment (neelanjan00) [inline chaoslib/litmus/container-kill/helper/container-kill.go:78]: Do we want to add the target information to the error as well?

File: chaoslib/litmus/disk-fill/helper/disk-fill.go [hunk 2]
```diff
@@ -119,63 +115,76 @@ func diskFill(experimentsDetails *experimentTypes.ExperimentDetails, clients cli
...
+			}
+
+			if err = result.AnnotateChaosResult(resultDetails.Name, chaosDetails.ChaosNamespace, "injected", "pod", t.Name); err != nil {
+				if remedyErr := remedy(t, clients); remedyErr != nil {
+					return remedyErr
+				}
+				return err
+			}
+		} else {
+			log.Warn("No required free space found!, It's Houseful")
```
Comment (neelanjan00) [inline chaoslib/litmus/disk-fill/helper/disk-fill.go:140]: It's Houseful? 😂

File: chaoslib/litmus/http-chaos/helper/http-helper.go [hunk 3]
```diff
@@ -67,66 +68,103 @@ func Helper(clients clients.ClientSets) {
...
+			continue
+		}
+		if err = result.AnnotateChaosResult(resultDetails.Name, chaosDetails.ChaosNamespace, "reverted", "pod", t.Name); err != nil {
+			errList = append(errList, err.Error())
+		}
 	}
 
-	return result.AnnotateChaosResult(resultDetails.Name, chaosDetails.ChaosNamespace, "reverted", "pod", experimentsDetails.TargetPods)
+	if len(errList) != 0 {
+		return fmt.Errorf(" failed to revert chaos, err: %v", strings.Join(errList, ","))
```
Comment (neelanjan00) [inline chaoslib/litmus/http-chaos/helper/http-helper.go:152]:
```suggestion
		return fmt.Errorf("failed to revert chaos, err: %v", strings.Join(errList, ","))
```

File: chaoslib/litmus/http-chaos/helper/http-helper.go [hunk 4]
```diff
@@ -67,66 +68,103 @@ func Helper(clients clients.ClientSets) {
...
+		types.SetEngineEventAttributes(eventsDetails, types.ChaosInject, msg, "Normal", chaosDetails)
+		events.GenerateEvents(eventsDetails, clients, chaosDetails, "ChaosEngine")
 	}
 
 	log.Infof("[Chaos]: Waiting for %vs", experimentsDetails.ChaosDuration)
 
 	common.WaitForDuration(experimentsDetails.ChaosDuration)
 
-	log.Info("[Chaos]: Stopping the experiment")
+	log.Info("[Chaos]: duration is over, reverting chaos")
```
Comment (neelanjan00) [inline chaoslib/litmus/http-chaos/helper/http-helper.go:136]:
```suggestion
	log.Info("[Chaos]: Chaos duration is over, reverting chaos")
```

File: chaoslib/litmus/network-chaos/lib/network-chaos.go [hunk 5]
```diff
@@ -460,9 +430,90 @@ func SetChaosTunables(experimentsDetails *experimentTypes.ExperimentDetails) {
...
+			"NetworkPacketLossPercentage": experimentsDetails.NetworkPacketLossPercentage,
+			"Sequence":                    experimentsDetails.Sequence,
+			"PodsAffectedPerc":            experimentsDetails.PodsAffectedPerc,
+		})
+	case "network-latency":
+		log.InfoWithValues("[Info]: The chaos tunables are:", logrus.Fields{
+			"NetworkLatency":   strconv.Itoa(experimentsDetails.NetworkLatency),
+			"Sequence":         experimentsDetails.Sequence,
+			"PodsAffectedPerc": experimentsDetails.PodsAffectedPerc,
+		})
```
Comment (neelanjan00) [inline chaoslib/litmus/network-chaos/lib/network-chaos.go:505]: We should also add Jitter here

File: chaoslib/litmus/pod-dns-chaos/helper/dnschaos.go [hunk 6]
```diff
@@ -80,77 +131,127 @@ func preparePodDNSChaos(experimentsDetails *experimentTypes.ExperimentDetails, c
...
+}
 
-	if err = result.AnnotateChaosResult(resultDetails.Name, chaosDetails.ChaosNamespace, "injected", "pod", experimentsDetails.TargetPods); err != nil {
-		return err
+func terminateProcess(cmd *exec.Cmd) error {
+	// kill command
+	killTemplate := fmt.Sprintf("sudo kill %d", cmd.Process.Pid)
+	kill := exec.Command("/bin/bash", "-c", killTemplate)
+	if err = kill.Run(); err != nil {
+		log.Errorf("unable to kill dns interceptor process cry, err :%v", err)
```
Comment (neelanjan00) [inline chaoslib/litmus/pod-dns-chaos/helper/dnschaos.go:218]: The cry emoji is wrongly formatted

File: chaoslib/litmus/stress-chaos/helper/stress-helper.go [hunk 7]
```diff
@@ -529,15 +558,18 @@ func abortWatcher(targetPID int, resultName, chaosNS, targetPodName string) {
...
 	for retry > 0 {
-		if err = terminateProcess(targetPID); err != nil {
-			log.Errorf("unable to kill stress process, err :%v", err)
+		for _, t := range targets {
+			if err = terminateProcess(t.Cmd.Process.Pid); err != nil {
+				log.Errorf("unable to revert for %v pod, err :%v", t.Name, err)
+				continue
+			}
+			if err = result.AnnotateChaosResult(resultName, chaosNS, "reverted", "pod", t.Name); err != nil {
+				log.Errorf("unable to annotate the chaosresult for %v pod, err :%v", t.Name, err)
```
Comment (neelanjan00) [inline chaoslib/litmus/stress-chaos/helper/stress-helper.go:567]:
```suggestion
				log.Errorf("[Abort]: Unable to annotate the chaosresult for %v pod, err :%v", t.Name, err)
```

File: chaoslib/litmus/stress-chaos/helper/stress-helper.go [hunk 8]
```diff
@@ -529,15 +558,18 @@ func abortWatcher(targetPID int, resultName, chaosNS, targetPodName string) {
 	// retry thrice for the chaos revert
 	retry := 3
 	for retry > 0 {
-		if err = terminateProcess(targetPID); err != nil {
-			log.Errorf("unable to kill stress process, err :%v", err)
+		for _, t := range targets {
+			if err = terminateProcess(t.Cmd.Process.Pid); err != nil {
+				log.Errorf("unable to revert for %v pod, err :%v", t.Name, err)
```
Comment (neelanjan00) [inline chaoslib/litmus/stress-chaos/helper/stress-helper.go:563]:
```suggestion
				log.Errorf("[Abort]: Unable to revert for %v pod, err :%v", t.Name, err)
```

File: pkg/utils/common/common.go [hunk 9]
```diff
@@ -83,25 +85,25 @@ func AbortWatcherWithoutExit(expname string, clients clients.ClientSets, resultD
 	// updating the chaosresult after stopped
 	failStep := "Chaos injection stopped!"
 	types.SetResultAfterCompletion(resultDetails, "Stopped", "Stopped", failStep)
-	result.ChaosResult(chaosDetails, clients, resultDetails, "EOT")
+	if err := result.ChaosResult(chaosDetails, clients, resultDetails, "EOT"); err != nil {
+		log.Errorf("[ABORT]: failed to update result, err: %v", err)
```
Comment (neelanjan00) [inline pkg/utils/common/common.go:89]:
```suggestion
		log.Errorf("[ABORT]: Failed to update result, err: %v", err)
```

File: pkg/utils/common/common.go [hunk 10]
```diff
@@ -83,25 +85,25 @@ func AbortWatcherWithoutExit(expname string, clients clients.ClientSets, resultD
...
+	}
+	log.Info("[ABORT]: updated chaosresult post stop")
```
Comment (neelanjan00) [inline pkg/utils/common/common.go:91]:
```suggestion
	log.Info("[ABORT]: Updated chaosresult post stop")
```

File: pkg/utils/common/common.go [hunk 11]
```diff
@@ -83,25 +85,25 @@ func AbortWatcherWithoutExit(expname string, clients clients.ClientSets, resultD
...
-}
-
-//GetIterations derive the iterations value from given parameters
-func GetIterations(duration, interval int) int {
-	var iterations int
-	if interval != 0 {
-		iterations = duration / interval
+	err = events.GenerateEvents(eventsDetails, clients, chaosDetails, "ChaosResult")
+	if err != nil {
+		log.Errorf("failed to create chaosresult abort event, err: %v", err)
```
Comment (neelanjan00) [inline pkg/utils/common/common.go:105]:
```suggestion
		log.Errorf("[ABORT]: Failed to create chaosresult abort event, err: %v", err)
```

File: chaoslib/litmus/disk-fill/helper/disk-fill.go (hunk 2 above)
Comment (avaakash) [reply chaoslib/litmus/disk-fill/helper/disk-fill.go:140]: The memory is partying XD

Comment (neelanjan00) [reply chaoslib/litmus/disk-fill/helper/disk-fill.go:140]: When disks have a better social life than you 🥲

File: chaoslib/litmus/http-chaos/helper/http-helper.go [hunk 12]
```diff
@@ -67,66 +68,103 @@ func Helper(clients clients.ClientSets) {
...
 	}
 
-	// record the event inside chaosengine
-	if experimentsDetails.EngineName != "" {
-		msg := "Injecting " + experimentsDetails.ExperimentName + " chaos on application pod"
-		types.SetEngineEventAttributes(eventsDetails, types.ChaosInject, msg, "Normal", chaosDetails)
-		events.GenerateEvents(eventsDetails, clients, chaosDetails, "ChaosEngine")
+	var targets []targetDetails
+
+	for _, t := range strings.Split(targetEnv, ";") {
```
Comment (avaakash) [inline chaoslib/litmus/http-chaos/helper/http-helper.go:78]: Can we convert this into a common function if the same code is used in every helper?

File: chaoslib/litmus/http-chaos/helper/http-helper.go [hunk 13]
```diff
@@ -67,66 +68,103 @@ func Helper(clients clients.ClientSets) {
...
-		// proceed for chaos injection
-		if err := startProxy(experimentDetails, pid); err != nil {
-			_ = killProxy(experimentDetails, pid)
-			return errors.Errorf("failed to start proxy, err: %v", err)
-		}
-		if err := addIPRuleSet(experimentDetails, pid); err != nil {
-			_ = killProxy(experimentDetails, pid)
-			return errors.Errorf("failed to add ip rule set, err: %v", err)
-		}
+	// proceed for chaos injection
```
Comment (avaakash) [inline chaoslib/litmus/http-chaos/helper/http-helper.go:160]:
```suggestion
```
this comment is no longer required

**Review by avaakash — CHANGES_REQUESTED** [review]

File: chaoslib/litmus/container-kill/helper/container-kill.go (hunk 1 above)
Comment (ispeakc0de) [reply chaoslib/litmus/container-kill/helper/container-kill.go:78]: It is killing all containers at once so if I will  append all container/pod names into error then error msg will become too long. Anyways we are logging {podName and targetContainer} duo for all pods/containers

**Review by ksatchit — APPROVED** [review]

**Review by uditgaurav — APPROVED** [review]

**Review by neelanjan00 — APPROVED** [review]

**Pull request merged**